README.md
# Operate API Client for Node.js

 [![NPM](https://nodei.co/npm/operate-api-client.png)](https://npmjs.org/package/operate-api-client) 

![Community Extension](https://img.shields.io/badge/Community%20Extension-An%20open%20source%20community%20maintained%20project-FF4700)

![Lifecycle](https://img.shields.io/badge/Lifecycle-Stable-brightgreen)

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

A Node.js client for interacting with the [Camunda 8 Operate REST API](https://docs.camunda.io/docs/apis-clients/operate-api/) in Camunda 8 SaaS.

Uses [camunda-saas-oauth-nodejs](https://github.com/camunda-community-hub/camunda-saas-oauth-nodejs) to use client credentials from the environment for authentication.

## Installation

```
npm i operate-api-client
```

## Usage

Set the credentials for Camunda SaaS in the environment, then:

```typescript
import { OperateApiClient } from 'operate-api-client'

const operate = new OperateApiClient()

operate.searchProcessInstances({
    filter: {
        state: "ACTIVE"
    },
    size: 50
}).then(instances => {
    console.log(instances)
})
```



CHANGELOG.md
# Change log

## Version 1.2.3

-   Add private method `safeJSONparse` to handle non-serializable variables.
-   In version 1.2.2, a maximum of 10 variables were returned by `getJSONVariablesforProcess`. In version 1.2.3, up to 1000 variables are returned.
## Version 1.2.2

### Enhancements

-   Add method `getJSONVariablesforProcess` to return variables as an object.


## Version 1.2.0

### Enhancements

-   Support custom OAuthProvider and baseUrl via constructor to enable multiple clients/clusters per application. See [#4](https://github.com/camunda-community-hub/operate-client-node-js/issues/4) for more details.


src/__test__/test.spec.ts
import { ProcessDefinition, ProcessInstance, Query } from "../lib/APIObjects";
import { OperateApiClient } from "../";

const c = new OperateApiClient();
jest.setTimeout(15000)

xtest("It can get the Incident", async () => {
    const res = await c.searchIncidents({
      filter: { 
        processInstanceKey: 2251799816400111
      }
    })
    console.log(JSON.stringify(res, null, 2))
    expect(res.total).toBe(1)
})
xtest("It can search process definitions", async () => {
  const query: Query<ProcessDefinition> = {
    filter: {},
    size: 50,
    sort: [
      {
        field: "bpmnProcessId",
        order: "ASC",
      },
    ],
  };
  const defs = await c.searchProcessDefinitions(query);
  expect(defs.total).toBeGreaterThanOrEqual(0);
});

xtest("It can get a specific process definition", async () => {
  const p = await c.getProcessDefinition(2251799817140074);
  expect(p).toBeTruthy();
});

xtest("It can get the process definition XML", async () => {
  const p = await c.getProcessDefinitionXML(2251799817140074);
  expect(p).toBeTruthy();
});

xtest("It can search for process instances", async () => {
  const query: Query<ProcessInstance>  = {
    filter: {
      processVersion: 1
    },
    size: 50,
    sort: [
      {
        field: "bpmnProcessId",
        order: "ASC",
      },
    ],
  };
  const defs = await c.searchProcessInstances(query);
  expect(defs).toBeTruthy();
  const d = await c.searchProcessInstances({});
  expect(d).toBeTruthy();
});

xtest("It can find a specific process instance", async () => {
  const query: Query<ProcessInstance> = {
    filter: {
      processVersion: 1,
      key: 2251799819847322,
    },
    size: 50,
    sort: [
      {
        field: "bpmnProcessId",
        order: "ASC",
      },
    ],
  };
  const defs = await c.searchProcessInstances(query);
  expect(defs).toBeTruthy();
});

xtest("It can get a specific process instance", async () => {
  const p = await c.getProcessInstance(2251799819847322);
  expect(p).toBeTruthy();
});

xtest("It can find incidents", async () => {
  const is = await c.searchIncidents();
  console.log(is)
  expect(is).toBeTruthy();
});

xtest("It can get a specific incident", async () => {
  const i = await c.getIncident(2251799818436725);
  expect(i).toBeTruthy();
});

xtest("It can get variables for a specific process", async () => {
  const vars = await c.getVariablesforProcess(4503599629029980)
  console.log(vars)
  expect(vars).toBeTruthy()
})

src/lib/APIObjects.ts
export interface ProcessDefinition {
  key: number;
  name: string;
  version: number;
  bpmnProcessId: string;
}

export interface ProcessInstance {
  key: number;
  processVersion: number;
  bpmnProcessId: string;
  parentKey?: number;
  /* yyyy-MM-dd'T'HH:mm:ss.SSSZZ */
  startDate: string;
  /* yyyy-MM-dd'T'HH:mm:ss.SSSZZ */
  endDate: string;
  state: 'ACTIVE' | 'COMPLETED' | 'CANCELED';
  processDefinitionKey: number;
}

export interface Incident {
  key: number;
  processDefinitionKey: number;
  processInstanceKey: number;
  type: 'UNSPECIFIED' | 'UNKNOWN' | 'IO_MAPPING_ERROR' | 'JOB_NO_RETRIES' | 'CONDITION_ERROR' | 'EXTRACT_VALUE_ERROR' | 'CALLED_ELEMENT_ERROR' | 'UNHANDLED_ERROR_EVENT' | 'MESSAGE_SIZE_EXCEEDED' | 'CALLED_DECISION_ERROR' | 'DECISION_EVALUATION_ERROR'
  message: string;
  /* yyyy-MM-dd'T'HH:mm:ss.SSSZZ */
  creationTime: string; //
  state: 'ACTIVE' | 'RESOLVED';
}

export interface FlownodeInstance {
  key: number;
  processInstanceKey: number;
  /* yyyy-MM-dd'T'HH:mm:ss.SSSZZ */
  startDate: string;
  /* yyyy-MM-dd'T'HH:mm:ss.SSSZZ */
  endDate: string;
  incidentKey: number;
  type:  'UNSPECIFIED' |
  'PROCESS' |
  'SUB_PROCESS' |
  'EVENT_SUB_PROCESS' |
  'START_EVENT' |
  'INTERMEDIATE_CATCH_EVENT' |
  'INTERMEDIATE_THROW_EVENT' |
  'BOUNDARY_EVENT' |
  'END_EVENT' |
  'SERVICE_TASK' |
  'RECEIVE_TASK' |
  'USER_TASK' |
  'MANUAL_TASK' |
  'TASK' |
  'EXCLUSIVE_GATEWAY' |
  'INCLUSIVE_GATEWAY' |
  'PARALLEL_GATEWAY' |
  'EVENT_BASED_GATEWAY' |
  'SEQUENCE_FLOW' |
  'MULTI_INSTANCE_BODY' |
  'CALL_ACTIVITY' |
  'BUSINESS_RULE_TASK' |
  'SCRIPT_TASK' |
  'SEND_TASK' |
  'UNKNOWN';
  state: 'ACTIVE' | 'COMPLETED' | 'TERMINATED';
  incident: boolean;
}

export interface Variables {
  key: number;
  processInstanceKey: number;
  scopeKey: number;
  name: string;
  /* Always truncated if value is too big in "search" results. In "get object" result it is not truncated. */
  value: string;
  /* if true 'value' is truncated. */
  truncated: boolean;
}

export interface ChangeStatus {
  /* What was changed */
  message: string;
  /* How many items were deleted */
  deleted: number;
}

export interface Query<T> {
  filter?: Partial<T>;
  size?: number;
  sort?: [{ field: keyof T; order: "ASC" | "DESC" }];
  searchAfter?: any[];
}

export interface SearchResults<T> {
  items: T[]
  sortValues: any[]
  total: number
}

src/lib/OperateClient.ts
import { OAuthProvider } from "camunda-saas-oauth";
import got from 'got';
import ky from 'ky-universal'
import { ChangeStatus, FlownodeInstance, Incident, ProcessDefinition, ProcessInstance, Query, SearchResults, Variables } from "./APIObjects";

const pkg = require('../../package.json')

const OPERATE_API_VERSION = 'v1'

export class OperateClient {
    oAuthProvider: OAuthProvider;
    userAgentString: string;
    baseUrl: string;
    httpOptions: { prefixUrl: string; };

    constructor({
        baseUrl = 'https://bru-2.operate.camunda.io', 
        oAuthProvider, 
        userAgentString
    }: 
    {
        baseUrl: string, 
        oAuthProvider: OAuthProvider,
        userAgentString: string
    }) {
        this.userAgentString = userAgentString
        this.oAuthProvider = oAuthProvider
        this.baseUrl = baseUrl
        const clusterId = this.oAuthProvider.zeebeAudience.split('.')[0]
        this.httpOptions = {
            prefixUrl: `${this.baseUrl}/${clusterId}/${OPERATE_API_VERSION}`
        }
    }

    private async getHeaders() {
        return {
            'content-type': 'application/json',
            'authorization': `Bearer ${await this.oAuthProvider.getToken('OPERATE')}`,
            'user-agent': this.userAgentString,
            'accept': '*/*'
        }
    }

    public async searchProcessDefinitions(query: Query<ProcessDefinition> = {}): Promise<SearchResults<ProcessDefinition>> {
        const headers = await this.getHeaders()
        return ky.post(`process-definitions/search`, {
            json: query,
            headers,
            ...this.httpOptions
        }).json()
    }

    public async getProcessDefinition(processDefinitionKey: number): Promise<ProcessDefinition> {
        const headers = await this.getHeaders()
        return ky(`process-definitions/${processDefinitionKey}`, {
            headers,
            ...this.httpOptions
        }).json()
    }

    public async getProcessDefinitionXML(processDefinitionKey: number): Promise<string> {
        const headers = await this.getHeaders()
        return ky(`process-definitions/${processDefinitionKey}/xml`, {
            headers,
            ...this.httpOptions
        }).text()
    }

    public async searchProcessInstances(query: Query<ProcessInstance> = {}): Promise<SearchResults<ProcessInstance>> {
        const headers = await this.getHeaders()
        return ky.post(`process-instances/search`, {
            json: query,
            headers,
            ...this.httpOptions
        }).json()
    }

    public async getProcessInstance(processInstanceKey: number): Promise<ProcessInstance> {
        const headers = await this.getHeaders()
        return ky(`process-instances/${processInstanceKey}`, {
            headers,
            ...this.httpOptions
        }).json()
    }

    public async deleteProcessInstance(processInstanceKey: number): Promise<ChangeStatus> {
        const headers = await this.getHeaders()
        return got.delete(`process-instances/${processInstanceKey}`, {
            headers,
            ...this.httpOptions
        }).json()
    }

    public async searchIncidents(query: Query<Incident> = {}): Promise<SearchResults<Incident>> {
        const headers = await this.getHeaders()
        return ky.post(`incidents/search`, {
            json: query,
            headers,
            ...this.httpOptions
        }).json()
    }

    public async getIncident(key: number): Promise<Incident> {
        const headers = await this.getHeaders()
        return got(`incidents/${key}`, {
            headers,
            ...this.httpOptions
        }).json()
    }

    public async searchFlownodeInstances(query: Query<FlownodeInstance>): Promise<SearchResults<FlownodeInstance>> {
        const headers = await this.getHeaders()
        return ky.post(`flownodes/search`, {
            headers,
            ...this.httpOptions,
            json: query
        }).json()
    }

    public async getFlownodeInstance(key: number): Promise<FlownodeInstance> {
        const headers = await this.getHeaders()
        return ky(`flownodes/${key}`, {
            headers,
            ...this.httpOptions
        }).json()        
    }

    public async searchVariables(query: Query<Variables>): Promise<SearchResults<Variables>> {
        const headers = await this.getHeaders()
        return ky.post(`variables/search`, {
            headers,
            json: query,
            ...this.httpOptions
        }).json()
    }

    public async getVariables(key: number): Promise<Variables> {
        const headers = await this.getHeaders()
        return ky(`variables/${key}`, {
            headers,
            ...this.httpOptions
        }).json()
    }
}

src/lib/OperateApiClient.ts
import { OAuthProviderImpl, getOperateToken } from "camunda-saas-oauth";
import { getOperateCredentials } from "camunda-8-credentials-from-env"
import got from 'got';
import { ChangeStatus, FlownodeInstance, Incident, ProcessDefinition, ProcessInstance, Query, SearchResults, Variable } from "./APIObjects";

const pkg = require('../../package.json')

const OPERATE_API_VERSION = 'v1'

type JSONDoc = { [key: string ]: (string | boolean | number | JSONDoc) }
/**
 * @description The high-level client for Operate.
 * @example
 * ```
 * const operate = new OperateApiClient()
 * 
 * operate.searchProcessInstances({
 *     filter: {
 *         state: "ACTIVE"
 *     },
 *     size: 50
 * }).then(instances => {
 *     console.log(instances)
 * })
 * ```
 */
export class OperateApiClient {
    private userAgentString: string;
    private gotOptions: { prefixUrl: string; };
    oauthProvider: OAuthProviderImpl | undefined;

    /**
     * @example
     * ```
     * const operate = new OperateApiClient()
     * ```
     */
    constructor(options: {
        oauthProvider?: OAuthProviderImpl,
        baseUrl?: string
    } = {}) {
        this.oauthProvider = options.oauthProvider
        this.userAgentString = `operate-client-nodejs/${pkg.version}`
        const baseUrl = options.baseUrl ?? getOperateCredentials().CAMUNDA_OPERATE_BASE_URL
        this.gotOptions = {
            prefixUrl: `${baseUrl}/${OPERATE_API_VERSION}`
        }
    }

    private async getHeaders() {
        const token = (this.oauthProvider) ? 
            await this.oauthProvider.getToken('OPERATE') : 
            await getOperateToken(this.userAgentString)
 
        return {
            'content-type': 'application/json',
            'authorization': `Bearer ${token}`,
            'user-agent': this.userAgentString,
            'accept': '*/*'
        }
    }

    /**
     * @description Search and retrieve process definitions. 
     * 
     * [Camunda 8 Documentation](https://docs.camunda.io/docs/apis-clients/operate-api/#process-definition)
     * @example
     * ```
     * const query: Query<ProcessDefinition> = {
     *   filter: {},
     *   size: 50,
     *   sort: [
     *     {
     *       field: "bpmnProcessId",
     *       order: "ASC",
     *     },
     *    ],
     *  };
     * const operate = newOperateClient()
     * const defs = await operate.searchProcessDefinitions(query);
     * ```
     */
    public async searchProcessDefinitions(query: Query<ProcessDefinition> = {}): Promise<SearchResults<ProcessDefinition>> {
        const headers = await this.getHeaders()
        return got.post(`process-definitions/search`, {
            json: query,
            headers,
            ...this.gotOptions
        }).json()
    }

    /**
     * 
     * @description Retrieve the metadata for a specific process definition, by key.  
     * 
     * [Camunda 8 Documentation](https://docs.camunda.io/docs/apis-clients/operate-api/#process-definition)
     * @example
     * ```
     * const operate = new OperateApiClient()
     * const definition = await operate.getProcessDefinition(2251799817140074);
     *  ```
     */
    public async getProcessDefinition(processDefinitionKey: number | string): Promise<ProcessDefinition> {
        const headers = await this.getHeaders()
        return got(`process-definitions/${processDefinitionKey}`, {
            headers,
            ...this.gotOptions
        }).json()
    }

    public async getProcessDefinitionXML(processDefinitionKey: number | string): Promise<string> {
        const headers = await this.getHeaders()
        return got(`process-definitions/${processDefinitionKey}/xml`, {
            headers,
            ...this.gotOptions
        }).text()
    }

    /**
     * @description Search and retrieve process instances.
     * @example 
     * ```
     * const operate = new OperateApiClient()
     * const query: Query<ProcessInstance>  = {
     *   filter: {
     *     processVersion: 1
     *   },
     *   size: 50,
     *   sort: [
     *     {
     *       field: "bpmProcessId",
     *       order: "ASC"
     *     }
     *   ]
     * }
     * const instances = await operate.searchProcessInstances(query)
     * console.log(`Found ${instances.total} instances`)
     */
    public async searchProcessInstances(query: Query<ProcessInstance> = {}): Promise<SearchResults<ProcessInstance>> {
        const headers = await this.getHeaders()
        return got.post(`process-instances/search`, {
            json: query,
            headers,
            ...this.gotOptions
        }).json()
    }

    /**
     * 
     * @description Retrieve a specific process instance by id. 
     * @example
     * ```
     * const operate = new OperateApiClient()
     * const instance = await operate.getProcessInstance(2251799819847322)
     * ```
     */
    public async getProcessInstance(processInstanceKey: number | string): Promise<ProcessInstance> {
        const headers = await this.getHeaders()
        return got(`process-instances/${processInstanceKey}`, {
            headers,
            ...this.gotOptions
        }).json()
    }

    /**
     * @description Delete a specific process instance by key.
     * @example  
     * ```
     * const operate = new OperateApiClient()
     * await operate.deleteProcessInstance(2251799819847322)
     * ``` 
     */
    public async deleteProcessInstance(processInstanceKey: number | string): Promise<ChangeStatus> {
        const headers = await this.getHeaders()
        return got.delete(`process-instances/${processInstanceKey}`, {
            headers,
            ...this.gotOptions
        }).json()
    }

    /**
     * @description Get the statistics for a process instance, grouped by flow nodes
     */
    public async getProcessInstanceStatistics(processInstanceKey: number | string): Promise<{
        // The id of the flow node for which the results are aggregated
        activityId: string,
        active: number,
        canceled: number,
        incidents: number,
        completed: number
    }[]> {
        const headers = await this.getHeaders()
        return got(`process-instances/${processInstanceKey}/statistics`, {
            headers,
            ...this.gotOptions
        }).json()
    }

    /**
     * @description Get sequence flows of process instance by key
     */
    public async getProcessInstanceSequenceFlows(processInstanceKey: number | string): Promise<string[]> {
        const headers = await this.getHeaders()
        return got(`process-instances/${processInstanceKey}/sequence-flows`, {
            headers,
            ...this.gotOptions
        }).json()        
    }

    /**
     * 
     * @description Search and retrieve incidents. 
     * @example
     * ```
     * const operate = new OperateApiClient()
     * const query: Query<Incident> = {
     *   filter: {
     *     state: "ACTIVE"
     *   },
     *   size: 50,
     *   sort: [
     *     {
     *       field: "creationTime",
     *       order: "ASC"
     *     }
     *   ]
     * }
     * const incidents = operate.searchIncidents(query)
     * ``` 
     */
    public async searchIncidents(query: Query<Incident> = {}): Promise<SearchResults<Incident>> {
        const headers = await this.getHeaders()
        return got.post(`incidents/search`, {
            json: query,
            headers,
            ...this.gotOptions
        }).json()
    }

    /**
     * 
     * @description Retrieve an incident by incident key. 
     * @example
     * ```
     * const operate = new OperateApiClient()
     * const incident = await operate.getIncident(2251799818436725)
     * console.log(incident.message)
     * ``` 
     */
    public async getIncident(key: number | string): Promise<Incident> {
        const headers = await this.getHeaders()
        return got(`incidents/${key}`, {
            headers,
            ...this.gotOptions
        }).json()
    }

    public async searchFlownodeInstances(query: Query<FlownodeInstance>): Promise<SearchResults<FlownodeInstance>> {
        const headers = await this.getHeaders()
        return got.post(`flownodes/search`, {
            headers,
            ...this.gotOptions,
            json: query
        }).json()
    }

    public async getFlownodeInstance(key: number | string): Promise<FlownodeInstance> {
        const headers = await this.getHeaders()
        return got(`flownodes/${key}`, {
            headers,
            ...this.gotOptions
        }).json()        
    }

    public async searchVariables(query: Query<Variable>): Promise<SearchResults<Variable>> {
        const headers = await this.getHeaders()
        return got.post(`variables/search`, {
            headers,
            json: query,
            ...this.gotOptions
        }).json()
    }

    /**
     * @description Retrieve the variables for a Process Instance, given its key
     * @param processInstanceKey 
     * @returns 
     */
    public async getVariablesforProcess(processInstanceKey: number | string): Promise<{items: Variable[]}> {
        const headers = await this.getHeaders()
        const body = {
            filter: {
                processInstanceKey
            }
        }
        return got.post(`variables/search`, {
            headers,
            body: JSON.stringify(body),
            ...this.gotOptions
        }).json()
    }

    /**
     * @description Retrieve the variables for a Process Instance as an object, given its key
     * @param processInstanceKey 
     * @returns 
     */
    public async getJSONVariablesforProcess<T extends {[key: string]: JSONDoc}>(processInstanceKey: number | string): Promise<T> {
        const headers = await this.getHeaders()
        const body = {
            filter: {
                processInstanceKey
            }
        }
        const vars: {items: Variable[]} = await got.post(`variables/search`, {
            headers,
            body: JSON.stringify(body),
            ...this.gotOptions
        }).json()
        return vars.items.reduce((prev, curr) => ({ ...prev, [curr.name]: JSON.parse(curr.value) }), {} as any)
    }

    /**
     * 
     * @description Return a variable identified by its variable key 
     * @returns 
     */
    public async getVariables(variableKey: number | string): Promise<Variable> {
        const headers = await this.getHeaders()
        return got(`variables/${variableKey}`, {
            headers,
            ...this.gotOptions
        }).json()
    }
}